src/Grpc.Net.Client/Internal/Retry/RetryCallBase.cs

```diff
@@ -0,0 +1,504 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Grpc.Net.Client.Internal.Http;
+using Grpc.Shared;
+using Microsoft.Extensions.Logging;
+
+#if NETSTANDARD2_0
+using ValueTask = System.Threading.Tasks.Task;
+#endif
+
+namespace Grpc.Net.Client.Internal.Retry
+{
+    internal abstract partial class RetryCallBase<TRequest, TResponse> : IGrpcCall<TRequest, TResponse>
+        where TRequest : class
+        where TResponse : class
+    {
+        private readonly TaskCompletionSource<IGrpcCall<TRequest, TResponse>> _commitedCallTcs;
+        private RetryCallBaseClientStreamReader<TRequest, TResponse>? _retryBaseClientStreamReader;
+        private RetryCallBaseClientStreamWriter<TRequest, TResponse>? _retryBaseClientStreamWriter;
+        private CancellationTokenRegistration? _ctsRegistration;
+
+        protected object Lock { get; } = new object();
```

Comment:
`Lock` is a bit too generic and doesn't convey what it's used to synchronize. Looking at the usages, maybe `BufferLock`, `WriteLock` or `OutputLock`?

Reply:
There is one lock for everything locking related. I didn't see the need to have more and introduce the possibility of deadlocks.